scripts/common.py

```diff
@@ -1023,3 +1032,286 @@ def managed_cloud_datastore_emulator(clear_datastore=False):
+        use_source_maps: bool. Whether to compile with source maps. Only
+            respected if config_path is None.
+        watch_mode: bool. Run the compiler in watch mode, which rebuilds on file
+            change.
+        max_old_space_size: int|None. Sets the max memory size of the compiler's
+            "old memory" section. As memory consumption approaches the limit,
+            the compiler will spend more time on garbage collection in an effort
+            to free unused memory.
+
+    Yields:
+        psutil.Process. The Webpack compiler process.
+    """
+    # TODO(#11549): Move this to top of the file.
+    import contextlib2
+
+    if config_path is not None:
+        pass
+    elif use_prod_env:
+        config_path = (
+            WEBPACK_PROD_SOURCE_MAPS_CONFIG if use_source_maps else
+            WEBPACK_PROD_CONFIG)
+    else:
+        config_path = (
+            WEBPACK_DEV_SOURCE_MAPS_CONFIG if use_source_maps else
+            WEBPACK_DEV_CONFIG)
+
+    compiler_args = [NODE_BIN_PATH, WEBPACK_BIN_PATH, '--config', config_path]
+    if max_old_space_size:
+        # NOTE: --max-old-space-size is a flag for Node.js, not the Webpack
+        # compiler, so we insert it immediately after NODE_BIN_PATH.
+        compiler_args.insert(1, '--max-old-space-size=%d' % max_old_space_size)
+    if watch_mode:
+        compiler_args.extend(['--color', '--watch', '--progress'])
+
+    with contextlib2.ExitStack() as exit_stack:
+        proc = exit_stack.enter_context(managed_process(
+            compiler_args, title='Webpack Compiler', shell=True,
+            # Capture compiler's output to detect when builds have completed.
+            stdout=subprocess.PIPE))
+
+        if watch_mode:
+            # Iterate until an empty string is printed, which signals the end of
+            # the process.
+            for line in iter(proc.stdout.readline, ''):
+                sys.stdout.write(line)
+                # Message printed when a compilation has succeeded. We break
+                # after the first one to ensure the site is ready to be visited.
+                if 'Built at: ' in line:
+                    break
+            else:
+                # If the code never ran `break`, raise an error because a build
+                # hasn't finished successfully.
+                raise IOError('First build never completed')
+
+        def print_proc_output():
+            """Prints the proc's output until it is exhausted."""
+            # Iterate until an empty string is printed, which signals the end of
+            # the output.
+            for line in iter(proc.stdout.readline, ''):
+                sys.stdout.write(line)
+
+        # Start a thread to print the rest of the compiler's output to stdout.
+        printer_thread = threading.Thread(target=print_proc_output)
+        printer_thread.start()
+        exit_stack.callback(printer_thread.join)
+
+        yield proc
+
+
+@contextlib.contextmanager
```

Comment:
(Optional) Given the number of these, and the fact that it's starting to become its own conceptual thing ... I wonder whether it's worth having a separate file in scripts/ that's just for managed processes, rather than sticking them all in common.py. What do you think?

Reply:
Done.